gpii/node_modules/flowManager/src/SettingsPutHandler.js

```diff
@@ -49,32 +50,37 @@ gpii.flowManager.cloudBased.settings.put.messages = {
 };
 
 gpii.flowManager.cloudBased.settings.put.handleRequest = function (that, gpiiKey, preferences) {
-    var authorizationPromise = gpii.oauth2.getAuthorization(that.req, that.authGrantFinder);
+    var accessToken = gpii.oauth2.parseAccessTokenFromRequest(that.req);
+    var authorizationPromise = gpii.oauth2.getAuthorization(accessToken, that.authGrantFinder);
 
     authorizationPromise.then(function (authorization) {
         if (authorization && authorization.gpiiKey === gpiiKey && authorization.allowSettingsPut) {
-            // TODO: Verify the received preferences with metadata to make sure they are the ones that are allowed to be updated.
-            // This is not supported at the first release of PSP
+            var prefsDetail = fluid.get(preferences, ["contexts", "gpii-default", "preferences"]);
+            if (authorization.allowedPrefsToWrite && Array.isArray(authorization.allowedPrefsToWrite) && !gpii.flowManager.cloudBased.settings.isPrefsAllowedToWrite(prefsDetail, authorization.allowedPrefsToWrite)) {
```

Comment:
It seems odd to place a check for allowedPrefsToWrite within the compound conditional here - this implies that if allowedPrefsToWrite is NOT an array, we just ignore its value and continue to allow the update. If we're not prepared to handle this case, we should reject the request.

Reply:
Yes, this is the intention as you said "if allowedPrefsToWrite is NOT an array, we just ignore its value and continue to allow the update", so that for client credentials used on public computers, [setting "allowedPrefsToWrite" to `null`](https://github.com/cindyli/universal/blob/GPII-3717/testData/dbData/clientCredentials.json#L10) will get the logic here ignore this validation and continue with the update. Of course, instead of setting it to `null`, we could set the array values here for all client credential documents. Each of these arrays has all Morphic supported preferences. But this means every time when Morphic supports new preferences, we'll need to perform a data migration to update these arrays to include the new preferences.

Please let me know if there's a better way to handle this case. Thanks.

Reply:
Yes, but what if the value is a non-array which is not null? If we find that allowedPrefsToWrite is, say, the string "Edward", we will still continue to allow everything. We should either signal an error or reject the request, or perhaps both.

Reply:
Great point. I will make the change.